brave/brave-ios
# Fix #8114: Support CoW Swap 🐮 orders in Safer Sign

## Summary of Changes
- Updates to use Safer Sign UI for displaying CoW swap orders.
- CoW Swap orders when swapping from ETH (native token) will create a `ETHSwap` transaction. For all other tokens, a sign message request is created (no gas fees!).
- A large part of this PR is refactoring UI components for re-use.
    - `SwapTransactionConfirmationView` was refactored into `SaferSignTransactionView` for re-use and removes dependency on `ParsedTransaction` for display. This was the safer sign UI for ethereum swap transactions.
        - `SaferSignTransactionContainerView` wraps this view for Safer Sign swap transactions.
        - `SaferSignMessageRequestContainerView` wraps this view for Safer Sign message requests (CoW Swap orders).
    - `SignMessageRequestView` was refactored to extract the content (message/domain display + consecutive new line & unknown character warnings). The message/domain & warning can now be displayed separately using `SignMessageRequestContentView`
        - `SaferSignMessageRequestContainerView` re-uses this `SignMessageRequestContentView` for the detail view, so we get the same display as sign message requests.
- A `SignMessageRequestStore` was created for `SignMessageRequestContainerView`. It's used to fetch details for CoW swap orders from core apis.

This pull request fixes #8114

## Submitter Checklist:


## Test Plan:

Direct link to CoW Swap for WETH token -> COW token on Goerli Test Network (I used Goerli, any supported network sufficient).
https://swap.cow.fi/#/5/swap/WETH/COW

1. Test a CoW Swap trade using any non-native token.
2. Verify:
    1. Safer Sign Swap UI is shown. 
    2. Tapping `Details` shows the sign transaction details. 
    3. Tapping link button beside tokens links out to explorer website for the token.
    4. Gas fee is free!
3. Test a CoW Swap trade with a custom recipient (tap the gear, enable Custom Recipient toggle).
4. Verify:
    1. Safer Sign Swap UI is shown
    2. Verify custom recipient is listed in the to token section (tap & hold to see full address).
    3. Links + gas fee verified above in Step 2.
5. Test a CoW Swap trade using native token (ex. ETH on Goerli).
    - ⚠️ These trades cost gas, since they are real transactions. Also, we do not have the ability to display custom recipients in this case, for the time being.
6. Verify Safer Sign Swap UI is shown as expected (similar to Swap in native Brave Wallet). Gas Fee shouldn't be 0, as this is a transaction instead of a signature request.
7. Test a CoW swap with unknown tokens. If you used COW token for any of the above, this has been verified by the COW symbol displaying. Any token not in search assets should suffice!


## Screenshots:

![CoW Swap](https://github.com/brave/brave-ios/assets/5314553/12a5912e-b9e8-4fe5-99db-20c9e648d3cb) | ![CoW Swap - details](https://github.com/brave/brave-ios/assets/5314553/07f3a5a4-9af5-4e14-b952-ce2adc1966d0) | _ 
--|--|--
![CoW Swap - custom recipient](https://github.com/brave/brave-ios/assets/5314553/a47f999c-3f6f-4c9a-82dd-4ca96b1fe1bd) | ![CoW Swap - custom recipient address](https://github.com/brave/brave-ios/assets/5314553/16b6e64f-8d97-44f5-a383-4f5e67ae30e7) | ![CoW Swap Details - custom recipient address](https://github.com/brave/brave-ios/assets/5314553/23efe180-55f1-452c-937e-4faa5787865e)

## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
